File: Old/thirtynineteen.py
#!/usr/bin/env python3
# Author: Sri Vangaru

import sys, resource

# resource.setrlimit(resource.RLIMIT_STACK, (-1, -1))
sys.setrecursionlimit(1000000)

# utility function for Kosaraju's algorithm
def dfs(node):
    visited[node] = True
    for next in graph[node]:
        if not visited[next]:
            dfs(next)
    reverse_topological.append(node)

# utility function for Kosaraju's algorithm
def dfs2(node, start):
    visited[node] = True
    scc[node] = start
    for next in transpose_graph[node]:
        if not visited[next]:
            dfs2(next, start)

# utility function for finding the reverse topological sorting of the
# condensed graph's nodes
def dfs3(condensed_node):
    visited[condensed_node] = True
    for next in condensed_graph[condensed_node]:
        if not visited[next]:
            dfs3(next)
    reverse_topological.append(condensed_node)

# sets up directed graph and transpose graph, and takes input
n, m = map(int, input().split())
graph = []
for i in range(n): graph.append([])
transpose_graph = []
for i in range(n): transpose_graph.append([])
for i in range(m):
    a, b = map(int, input().split())
    a -= 1
    b -= 1
    graph[a].append(b)
    transpose_graph[b].append(a)

# uses Kosaraju's algorithm to map each node to its strongly connected
# component (SCC), where each SCC is denoted by a representative node
visited = [False] * n
reverse_topological = []
for i in range(n):
    if not visited[i]: dfs(i)
visited = [False] * n
scc = [0] * n
for node in reversed(reverse_topological):
    if not visited[node]: dfs2(node, node)

# condenses the graph so that each SCC is treated as a singular node--keeping
# track of the original nodes within each SCC--resulting in a directed acylic
# graph (DAG)
condensed_graph = []
for i in range(n): condensed_graph.append([])
nodes_in_scc = []
for i in range(n): nodes_in_scc.append([])
for i in range(n):
    for adjacent in graph[i]:
        if scc[i] != scc[adjacent]:
            condensed_graph[scc[i]].append(scc[adjacent])
    nodes_in_scc[scc[i]].append(i)

# finds the reverse topological ordering (or normal DFS postorder ordering) of
# the nodes in the condensed graph
visited = [False] * n
reverse_topological.clear()
for i in range(n):
    if not visited[i] and i == scc[i]: dfs3(i)

# uses dynamic programming to find the longest path in the condensed graph,
# where the length of a path is defined by the number of bases (nodes from the
# original graph) covered in the path; this approach is allowed because the
# drone is allowed to visit the same base multiple times in its "path", so it
# can go around and around in circles within an SCC in order to land at the
# correct node in the SCC that has an outgoing edge to a different SCC, and
# similarly, any node within the start SCC is a valid start node, so we simply
# arbitrarily choose the SCC's representative node
memo = [0] * n
next = [-1] * n
for node in reverse_topological:
    max = 0
    for adjacent in condensed_graph[node]:
        if memo[adjacent] > max:
            max = memo[adjacent]
            next[node] = adjacent
    memo[node] = len(nodes_in_scc[node]) + max

# finds the node in the condensed graph that is the start of the longest path
max = -1
start = -1
for node in reverse_topological:
    if memo[node] > max:
        max = memo[node]
        start = node

# retrieves all of the nodes reachable within the longest path defined by the
# DP earlier
on_path = []
current_node = start
while True:
    for original_node in nodes_in_scc[current_node]:
        on_path.append(original_node)

    if next[current_node] == -1: break
    else: current_node = next[current_node]

# outputs the results
print(max)
for node in on_path: print(f"{node+1} ", end="")
print(f"\n{start+1}")

File: Old/ThirtyNineteen.java
// Author: Sri Vangaru

import java.util.*;
import java.io.*;

public class ThirtyNineteen {
    static ArrayList<Integer>[] graph;
    static ArrayList<Integer>[] transposeGraph;
    static ArrayList<Integer> reverseTopological;
    static boolean[] visited;
    static int[] scc;
    static TreeMap<Integer, TreeSet<Integer>> nodesInSCC;
    static ArrayList<Integer>[] condensed;
    static TreeSet<Integer> reachableOnPath;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        PrintWriter pw = new PrintWriter(System.out);

        // The problem can simply be stated as: find the (not necessarily unique)
        // longest path in a directed graph, where the path does not have to be
        // simple (i.e. it can repeat vertices) and the length of the path is
        // defined as the number of vertices within it.

        // Sets up directed graph and transpose graph, and then takes input
        String[] line = br.readLine().split(" ");
        int n = Integer.parseInt(line[0]);
        int m = Integer.parseInt(line[1]);
        graph = new ArrayList[n]; transposeGraph = new ArrayList[n];
        for (int i = 0; i <= n-1; i++) {
            graph[i] = new ArrayList<>();
            transposeGraph[i] = new ArrayList<>();
        }
        for (int i = 1; i <= m; i++) {
            line = br.readLine().split(" ");
            int a = Integer.parseInt(line[0]) - 1;
            int b = Integer.parseInt(line[1]) - 1;
            graph[a].add(b);
            transposeGraph[b].add(a);
        }

        // Uses Kosaraju's algorithm to map each node to its strongly connected component (SCC),
        // where each SCC is denoted by a representative node
        visited = new boolean[n];
        reverseTopological = new ArrayList<Integer>();
        for (int i = 0; i <= n-1; i++) if (!visited[i]) dfs(i);
        visited = new boolean[n];
        scc = new int[n];
        for (int i = n-1; i >= 0; i--) {
            int node = reverseTopological.get(i);
            if (!visited[node]) dfs2(node, node);
        }

        // Condenses the graph so that each SCC is treated as a singular node--keeping track
        // of the original nodes within each SCC--resulting in a directed acylic graph (DAG)
        condensed = new ArrayList[n];
        for (int i = 0; i <= n-1; i++) condensed[i] = new ArrayList<Integer>();
        nodesInSCC = new TreeMap<Integer, TreeSet<Integer>>();
        for (int i = 0; i <= n-1; i++) {
            for (int adjacent : graph[i]) {
                if (scc[i] != scc[adjacent]) {
                    condensed[scc[i]].add(scc[adjacent]);
                }
            }

            if (!nodesInSCC.containsKey(scc[i])) nodesInSCC.put(scc[i], new TreeSet<>());
            nodesInSCC.get(scc[i]).add(i);
        }

        // Finds the reverse topological ordering (or normal DFS postorder ordering) of the nodes
        // in the condensed graph
        visited = new boolean[n];
        reverseTopological = new ArrayList<Integer>();
        for (int i = 0; i <= n-1; i++) if (!visited[i] && i == scc[i]) dfs3(i);

        // Uses dynamic programming to find the longest path in the condensed graph, where
        // the length of a path is defined by the number of bases (nodes from the original
        // graph) covered in the path. This approach is allowed because the drone is allowed
        // to visit the same base multiple times in its "path", so it can go around and around
        // in circles within an SCC in order to land at the correct node in the SCC that has
        // an outgoing edge to a different SCC. Similarly, any node within the start SCC
        // is a valid start node, so we simply arbitrarily choose the SCC's representative
        // node.
        int max = Integer.MIN_VALUE, start = -1;
        int[] memo = new int[n];
        int[] next = new int[n]; Arrays.fill(next, -1);
        for (int i = 0; i <= reverseTopological.size()-1; i++) {
            int node = reverseTopological.get(i);
            int maxFromAdj = 0;
            for (int adjacent : condensed[node]) {
                if (memo[adjacent] > maxFromAdj) {
                    maxFromAdj = memo[adjacent];
                    next[node] = adjacent;
                }
            }
            memo[node] = maxFromAdj + nodesInSCC.get(node).size();

            if (memo[node] > max) {
                max = memo[node];
                start = node;
            }
        }

        // Retrieves all of the nodes reachable within the longest path defined
        // by the DP earlier
        reachableOnPath = new TreeSet<Integer>();
        int currentNode = start;
        while (true) {
            for (int nodeInSCC : nodesInSCC.get(currentNode)) reachableOnPath.add(nodeInSCC);

            if (next[currentNode] == -1) break;
            else currentNode = next[currentNode];
        }

        // Outputs the results
        pw.println(max);
        for (int node : reachableOnPath) pw.print(node+1+" ");
        pw.println("\n"+(start+1));

        br.close(); pw.close();
    }

    // Utility function for Kosaraju's algorithm
    static void dfs(int node) {
        visited[node] = true;
        for (int next : graph[node])
            if (!visited[next])
                dfs(next);
        
        reverseTopological.add(node);
    }

    // Utility function for Kosaraju's algorithm
    static void dfs2(int node, int start) {
        visited[node] = true;
        scc[node] = start;
        for (int next : transposeGraph[node])
            if (!visited[next])
                dfs2(next, start);
    }

    // Utility function for finding the reverse topological sorting of the
    // condensed graph's nodes
    static void dfs3(int node) {
        visited[node] = true;
        for (int next : condensed[node])
            if (!visited[next])
                dfs3(next);

        reverseTopological.add(node);
    }
}
